modules/mod-deep-learning/ModelCard.h

```diff
@@ -0,0 +1,212 @@
+#include <wx/string.h>
+#include <wx/log.h>
+
+#include "MemoryX.h"
+#include "AudacityException.h"
+
+using Doc = rapidjson::Document;
+
+// this exception should be caught internally, but we 
+// derive from MessageBoxException just in case it needs to 
+// get handled by Audacity
+class InvalidModelCardDocument final : public MessageBoxException
+{
+public:
+   InvalidModelCardDocument(const TranslatableString msg, std::string trace): 
+                           m_msg(msg), 
+                           m_trace(std::move(trace)),
+                            MessageBoxException {
+                               ExceptionType::Internal,
+                               XO("Invalid Model Card Document")
+                            }
+   { 
+   }
+
+   // detailed internal error message
+   virtual const char* what() const throw () 
+   {
+      wxLogError(m_msg.Translation());
+      if (!m_trace.empty()) 
+         wxLogError(wxString(m_trace)); 
+      // TODO: also print the document
+      // TODO: check if document is nullptr
+      return m_msg.Translation().c_str();
+   }
+
+   // user facing message
+   virtual TranslatableString ErrorMessage() const
+      { return XO("Model Card Error: %s")
+               .Format(m_msg);}
+
+   const TranslatableString m_msg;
+   const std::string m_trace;
+};
+
+namespace parsers 
+{
+   Doc ParseString(const std::string &json);
+   Doc ParseFile(const std::string &path);
+}
+
+// validators that accept a default value are no-throw
+namespace validators
+{
+
+   std::string tryGetString(const std::string &key, const Doc& doc);
+   std::string tryGetString(const std::string &key, const Doc& doc, const std::string &defaultValue);
+
+   int tryGetInt(const std::string &key, const Doc& doc);
+   int tryGetInt(const std::string &key, const Doc& doc, int defaultValue);
+
+   std::vector<std::string> tryGetStringArray(const std::string &key, const Doc& doc, bool lower=false);
+                                             
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc);
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc, uint64_t defaultValue);
+
+}
+
+class DeepModelManager;
+
+class ModelCard final
+{
+public: 
+   ModelCard() = default;
+   ModelCard(const ModelCard&) = default;
+
+   // returns {author}/{name}
+   std::string GetRepoID() const;
+
+   bool IsSame(const ModelCard& other) const { return (*this) == other; };
+
+private:
+   friend class DeepModelManager;
+   
+   // throws InvalidModelCardDocument if the given json is not valid. 
+   void DeserializeFromFile(const std::string& path, const Doc& schema);
+   void SerializeToFile(const std::string& path) const;
+
+   template < typename Writer >
+   void Serialize(Writer &writer) const;
+   void Deserialize(const Doc& doc, const Doc& schema);
+
+   void Validate(const Doc& doc, const Doc& schema);
+
+   bool IsLocal() { return m_is_local; }
+   void SetLocal(bool local) { m_is_local = local; }
```

Comment:
I checked again to see which of the "setter" member functions of ModelCard are used.

long_description, short_description, and labels are never used (where I think we might want to solve the i18n problem).

Also sample_rate, multichannel, effect_type, domain_tags and tags are not used.

Do you need to repair these omissions?

Reply:
mmmm, I think those setters are not used because those fields are filled out in `ModelCard::Deserialize` always (since they come from json data), but I wanted to leave them there for completeness. Any thoughts?